hypothesis-python/tests/cover/test_subnormal_floats.py

```diff
@@ -0,0 +1,114 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2021 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from sys import float_info
+
+import pytest
+
+from hypothesis.errors import InvalidArgument
+from hypothesis.internal.floats import next_down, next_up
+from hypothesis.strategies import floats
+from hypothesis.strategies._internal.numbers import next_down_normal, next_up_normal
+
+from tests.common.debug import assert_no_examples, find_any
+
+# See tests/nocover/test_subnormal_floats.py for my background
+pytestmark = [
+    pytest.mark.skipif(
+        next_down(float_info.min) == 0.0, reason="broken by unsafe compiler flags"
+    )
+]
+
+
+def kw(marks=(), **kwargs):
+    id_ = ", ".join(f"{k}={v!r}" for k, v in kwargs.items())
+    return pytest.param(kwargs, id=id_, marks=marks)
+
+
+@pytest.mark.parametrize(
+    "kwargs",
+    [
+        kw(min_value=1),
+        kw(min_value=1),
+        kw(max_value=-1),
+        kw(min_value=float_info.min),
+        kw(min_value=next_down(float_info.min), exclude_min=True),
+        kw(max_value=-float_info.min),
+        kw(min_value=next_up(-float_info.min), exclude_max=True),
+    ],
+)
+def test_subnormal_validation(kwargs):
+    strat = floats(**kwargs, allow_subnormal=True)
+    with pytest.raises(InvalidArgument):
+        strat.example()
+
+
+@pytest.mark.parametrize(
+    "kwargs",
+    [
+        # min value
+        kw(allow_subnormal=False, min_value=1),
+        kw(allow_subnormal=False, min_value=float_info.min),
+        kw(allow_subnormal=True, min_value=-1),
+        kw(allow_subnormal=True, min_value=next_down(float_info.min)),
+        # max value
+        kw(allow_subnormal=False, max_value=-1),
+        kw(allow_subnormal=False, max_value=-float_info.min),
+        kw(allow_subnormal=True, max_value=1),
+        kw(allow_subnormal=True, max_value=next_up(-float_info.min)),
+        # min/max values
+        kw(
+            allow_subnormal=True,
+            min_value=-1,
+            max_value=1,
+            marks=pytest.mark.xfail(
+                reason="FixedBoundFloatStrategy(0, 1) rarely generates subnormals"
```

Comment:
Let's make this a `skip` to save time, and to avoid strict-xfail making our tests fail if it passes.